Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Github5623.xaml.cs

```diff
@@ -0,0 +1,199 @@
+		static SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1, 1);
+
+		public Github5623()
+		{
+#if APP
+			Device.SetFlags(new List<string> { CollectionView.CollectionViewExperimental });
+
+			InitializeComponent();
+
+			BindingContext = new ViewModel5623();
+#endif
+		}
+
+		protected override void Init()
+		{
+
+		}
+
+		async void CollectionView_RemainingItemsThresholdReached(object sender, System.EventArgs e)
+		{
+			await SemaphoreSlim.WaitAsync();
+			try
+			{
+				var itemsSource = (sender as CollectionView).ItemsSource as ObservableCollection<Model5623>;
+
+				if (itemsSource.Count == _maximumItemCount)
+				{
+					System.Diagnostics.Debug.WriteLine("Count: " + itemsSource.Count);
+					return;
+				}
+
+				var nextSet = await GetNextSetAsync();
+
+				Device.BeginInvokeOnMainThread(() =>
+				{
+					foreach (var item in nextSet)
+					{
+						itemsSource.Add(item);
+					}
+				});
+
+				System.Diagnostics.Debug.WriteLine("Count: " + itemsSource.Count);
+			}
+			finally
+			{
+				SemaphoreSlim.Release();
+			}
+		}
+
+		void CollectionView_OnScrolled(object sender, Core.Items.ScrolledEventArgs e)
+		{
+			Label1.Text = "HorizontalDelta: " + e.HorizontalDelta;
+			Label2.Text = "VerticalDelta: " + e.VerticalDelta;
+			Label3.Text = "HorizontalOffset: " + e.HorizontalOffset;
+			Label4.Text = "VerticalOffset: " + e.VerticalOffset;
+			Label5.Text = "FirstVisibleItemIndex: " + e.FirstVisibleItemIndex;
+			Label6.Text = "LastVisibleItemIndex: " + e.LastVisibleItemIndex;
+		}
+
+		async Task<ObservableCollection<Model5623>> GetNextSetAsync()
+		{
+			return await Task.Run(() =>
+			{
+				var collection = new ObservableCollection<Model5623>();
+
+				for (var i = _itemCount; i < _itemCount + _pageSize; i++)
+				{
+					collection.Add(new Model5623((BindingContext as ViewModel5623).ItemSizingStrategy == ItemSizingStrategy.MeasureAllItems)
+					{
+						Text = i.ToString(),
+						BackgroundColor = i % 2 == 0 ? Color.AntiqueWhite : Color.Lavender
+					});
+				}
+
+				_itemCount += _pageSize;
+
+				return collection;
+			});
+		}
+
+#if UITEST
+		[Test]
+		public void CollectionViewInfiniteScroll()
+		{
+			RunningApp.WaitForElement (q => q.Marked ("CollectionView"));
+			
+			var elementQuery = RunningApp.Query(c => c.Marked((_maximumItemCount - 1).ToString()));
+			var elementAvailable = elementQuery.Any();
+			var attempt = 0;
+
+			while (!elementAvailable)
+			{
+				RunningApp.ScrollDown();
+
+				elementAvailable = elementQuery.Any();
+				if (elementAvailable)
+					break;
+
+				++attempt;
+
+				if (attempt == 30)
+				{
+					Assert.Fail("Failed to find the last element");
+					break;
+				}
+			}
+		}
+#endif
+	}
+
+	[Preserve(AllMembers = true)]
+	public class ViewModel5623
+	{
+		public ObservableCollection<Model5623> Items { get; set; }
+
+		public Command RemainingItemsThresholdReachedCommand { get; set; }
+
+		public ItemSizingStrategy ItemSizingStrategy { get; set; } = ItemSizingStrategy.MeasureAllItems;
```

Comment:
This strategy doesn't work well on iOS (see the first item as it isn't measured/laid out properly). Not sure if this is a known bug. When testing on iOS, you might want to change this to first item measure so every row gets the same size.